File: frontend/src/lib/types/system-backup.ts
import type { BackupDestination, BackupManagementType, BackupPolicy, BackupRun } from './backup';

export type SystemBackupDestination = BackupDestination;

export type SystemBackupRun = BackupRun;

export type BackupHistoryEntry = BackupRun & {
	type: BackupManagementType;
	resourceType: 'system' | 'volume';
	resourceName: string;
};

export type SystemVolumeBackupSelectionMode = 'all' | 'allowlist' | 'blocklist';

export type SystemVolumeBackupPolicy = BackupPolicy & {
	s3DestinationName?: string;
	selectionMode: SystemVolumeBackupSelectionMode;
	volumeNames: string[];
	ignoreAnonymous: boolean;
	lastRun?: SystemBackupRun;
};

export type UpdateSystemVolumeBackupPolicy = Omit<SystemVolumeBackupPolicy, 's3DestinationName' | 'lastRun'> & {
	s3DestinationId: string;
};

export type SystemVolumeBackupPolicyCollection = {
	policies: SystemVolumeBackupPolicy[];
};

export type SystemVolumeBackupOption = {
	name: string;
	anonymous: boolean;
	available: boolean;
};

export type SystemVolumeBackupRunResult = {
	activityId: string;
	status: 'running';
};

export type SystemVolumeBackupCustomRun = {
	destination: BackupDestination;
	s3DestinationId?: string;
	stopContainers: boolean;
	selectionMode: SystemVolumeBackupSelectionMode;
	volumeNames: string[];
	ignoreAnonymous: boolean;
};

export type RunSystemVolumeBackups = {
	policyId?: string;
	custom?: SystemVolumeBackupCustomRun;
};

export type SystemBackupPolicy = {
	id: string;
	enabled: boolean;
	schedule: string;
	retentionCount: number;
	localEnabled: boolean;
	s3Enabled: boolean;
	s3DestinationId?: string;
	s3DestinationName?: string;
	lastRun?: SystemBackupRun;
};

export type UpdateSystemBackupPolicy = Omit<SystemBackupPolicy, 's3DestinationName' | 'lastRun'>;

export type SystemBackupPolicyCollection = {
	policies: SystemBackupPolicy[];
	recoveryKeyStored: boolean;
};

export type SystemBackupRecoveryKey = {
	recoveryKey: string;
};

export type CreateSystemBackup = {
	destination?: SystemBackupDestination;
	s3DestinationId?: string;
	recoveryKey?: string;
	policyId?: string;
};

File: types/backup/system_backup.go
package backup

import "time"

type SystemBackupDestination string

const (
	SystemBackupDestinationLocal   SystemBackupDestination = "local"
	SystemBackupDestinationS3      SystemBackupDestination = "s3"
	SystemBackupDestinationLocalS3 SystemBackupDestination = "local_s3"
)

type SystemBackupRun struct {
	ID                string                  `json:"id"`
	Size              int64                   `json:"size"`
	CreatedAt         time.Time               `json:"createdAt"`
	Status            string                  `json:"status"`
	Trigger           string                  `json:"trigger"`
	Destination       SystemBackupDestination `json:"destination"`
	LocalSnapshotID   string                  `json:"localSnapshotId,omitempty"`
	RemoteSnapshotID  string                  `json:"remoteSnapshotId,omitempty"`
	S3DestinationID   string                  `json:"s3DestinationId,omitempty"`
	S3DestinationName string                  `json:"s3DestinationName,omitempty"`
	PolicyID          string                  `json:"policyId,omitempty"`
	Error             string                  `json:"error,omitempty"`
}

type SystemBackupPolicy struct {
	ID                string           `json:"id"`
	Enabled           bool             `json:"enabled"`
	Schedule          string           `json:"schedule"`
	RetentionCount    int              `json:"retentionCount"`
	LocalEnabled      bool             `json:"localEnabled"`
	S3Enabled         bool             `json:"s3Enabled"`
	S3DestinationID   string           `json:"s3DestinationId,omitempty"`
	S3DestinationName string           `json:"s3DestinationName,omitempty"`
	LastRun           *SystemBackupRun `json:"lastRun,omitempty"`
}

type UpdateSystemBackupPolicy struct {
	ID              string `json:"id,omitempty"`
	Enabled         bool   `json:"enabled"`
	Schedule        string `json:"schedule"`
	RetentionCount  int    `json:"retentionCount"`
	LocalEnabled    bool   `json:"localEnabled"`
	S3Enabled       bool   `json:"s3Enabled"`
	S3DestinationID string `json:"s3DestinationId,omitempty"`
}

type SystemBackupPolicyCollection struct {
	Policies          []SystemBackupPolicy `json:"policies"`
	RecoveryKeyStored bool                 `json:"recoveryKeyStored"`
}

type UpdateSystemBackupPolicies struct {
	Policies []UpdateSystemBackupPolicy `json:"policies"`
}

type SystemBackupRecoveryKeyStatus struct {
	Configured bool `json:"configured"`
}

type SystemBackupRecoveryKey struct {
	RecoveryKey string `json:"recoveryKey"`
}

type CreateSystemBackupRequest struct {
	Destination     SystemBackupDestination `json:"destination,omitempty"`
	S3DestinationID string                  `json:"s3DestinationId,omitempty"`
	RecoveryKey     string                  `json:"recoveryKey,omitempty"`
	PolicyID        string                  `json:"policyId,omitempty"`
}

type RestoreSystemBackupRequest struct {
	RecoveryKey string `json:"recoveryKey"`
}

type UploadSystemBackupRequest struct {
	S3DestinationID string `json:"s3DestinationId"`
	RecoveryKey     string `json:"recoveryKey"`
}

type DeleteSystemBackupRequest struct {
	RecoveryKey string `json:"recoveryKey,omitempty"`
}

type DiscoverSystemBackupsRequest struct {
	S3DestinationID string `json:"s3DestinationId"`
	RecoveryKey     string `json:"recoveryKey"`
}
